### PR TITLE
Add kv_namespaces in metadata

### LINKED ISSUE #92: `wrangler deploy` un-does KV namespace bindings
In three screenshots:

image

image

image

### PR DESCRIPTION
**Pull request by xtuc**
Fixes #92

### CODE REVIEW

File: src/commands/build/mod.rs [hunk 1]
```diff
@@ -45,8 +45,13 @@ pub fn build(cache: &Cache, project_type: &ProjectType) -> Result<(), failure::E
                 failure::bail!("Webpack returned an error");
             }
 
+            let mut kv_namespaces = vec![];
+            if let Some(value) = &project.kv_namespaces {
+                kv_namespaces = value.clone();
+            }
```
Comment (steveklabnik) [inline src/commands/build/mod.rs]: This whole block is more idiomatically written as:

    let kv_namespaces = project.kv_namespaces.unwrap_or(Vec::new());

(You may need a `.clone()` before the `.unwrap_or`, I'm not actually 100% sure)

**Comment by xtuc** [comment]
It kinda works, you need to specifiy the namespace_id in your wrangler.toml.

```toml
# ...

[[kv-namespaces]]
local_binding = "local_binding"
name = "name"
namespace_id = "abc"
```

**Comment by signalnerve** [comment]
Re: the `namespace_id` comment - I agree that having to specify it inside of `wrangler.toml` isn't great, but if users are creating the bindings from inside of a Wrangler-enabled project, can we write it to `wrangler.toml` for them on initial creation? 

There may be situations where they'll need to go look up the namespace manually, but I imagine the average workflow will take place (hopefully!) in Wrangler, so we should know the `namespace_id` of a KV store since we're creating it for them during the publishing process.

**Comment by steveklabnik** [comment]
What is currently holding this PR back from landing?

**Comment by xtuc** [comment]
@steveklabnik it would require a couple of rebases but I'm happy to see this landing. If you need it I can continue to work on it.

**Comment by steveklabnik** [comment]
It'd be nice to get it landed so that we can continue the work on KV support overall; if you've got the time that would be great.

**Comment by ashleymichal** [comment]
I think what I'm missing to be able to judge this PR is the README section on how to configure your namespaces in the wrangler.toml; @xtuc touched on it in his comment but what would my wrangler.toml look like if I had two or more namespaces exposed to my worker?

**Comment by steveklabnik** [comment]
@ashleymichal it currently looks like this:

```toml
kv-namespaces = ["STATIC_CONTENT"]
```

if you had two

```toml
kv-namespaces = ["STATIC_CONTENT", "FOOBAR"]

```

**Comment by ashleymichal** [comment]
@steveklabnik based on the earlier comment from @xtuc [here](https://github.com/cloudflare/wrangler/pull/215#issuecomment-500426900), that does not appear to be the case in the context of this PR; the user is required to specify not only the name of the binding as it should appear in the Worker, but the "title" and id of the namespace, in order to both create the namespace (if it does not already exist), and specify the binding in the metadata.

The question still stands, as once we provide any interface to users, it becomes difficult to change. In order to land this, we need to care about a few things:

1. Providing enough information to allow Wrangler to provision a KV namespace on behalf of a user OR requiring the user to provision the namespace before integrating with their Worker project.
2. Providing enough information for Wrangler to generate a KV Binding entry for the metadata part of the upload form. This can either be provided by the user manually (asking them to enter their namespace id manually as shown in @xtuc 's comment), or it can be retrieved by Wrangler behind the scenes with a query to the KV API based on the namespace title.
3. Providing configuration of the relationship between the binding variable name and the namespace title. This can be done as it stands currently (by assuming they are one and the same and requiring they adhere to JavaScript variable name standards), or it can be done more manually as specified in @xtuc 's comment.

I think 1. is well handled here and lays a decent enough foundation for later improvements and features.

In the case of 2 and 3, I think it would be short-sighted to go with the seemingly "simpler" approach of requiring the user to only provide the namespace title. It is much easier in the long term to make those three connections explicitly and provide "sane defaults" later than to do the reverse. We should therefore take the time to consider what the entries should look like in the wrangler.toml now.

**Comment by ashleygwilliams** [comment]
config wise- re multiple kv name spaces i'd probably do something like this... keying on name

```
[kv-namespaces]
"name1": {"local_binding": "thing", "id" : "aksdhjfkahg;a"}

[kv-namespaces.name1]
local_binding = "thing"
id = "aksdhjfkahg;a"
```

**Comment by ashleygwilliams** [comment]
it should also be noted this is only for the webpack project type which is a deliberate decision we should make

**Comment by steveklabnik** [comment]
Ah, see when I implemented it, it used the title of the namespace and made a binding of the same name; we decided not to support the case where the two differ.

**Comment by xtuc** [comment]
Maybe we could iterate on this PR, to get something useful I would:
- remove the auto creation of namespaces (because we don't store the id). The user/we should do it from the UI.
- let the namespace_id, name and binding name be in the wrangler.toml

This will unlock our deployement and provide a base used for future iterations. We shouldn't document it for now tho. What do you think?

**Comment by cwndrws** [comment]
I agree that the KV config should be in `wrangler.toml`. I like the idea of requiring all three values: (title, id, and binding) like @ashleymichal suggested. I think requiring title be a valid JS symbol would bite us eventually.

File: src/commands/build/wranglerjs/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+use serde::Serialize;
+
+#[derive(Serialize, Debug)]
+pub struct Bindings {
```
Comment (cwndrws) [inline src/commands/build/wranglerjs/metadata.rs]: We should make this generic and move out of webpack specific project type. Each binding type has different configuration values. This is what I was referring to in #273. We don't need to add all possible binding types, but make it possible add them in the future and make it possible to use the same binding type for both KV namespaces and wasm modules.

Comment (ashleymichal) [reply src/commands/build/wranglerjs/metadata.rs]: i need to brush up on my rust type system, but i think maybe binding is a trait... #hottake. gonna write this into the design doc i think.

Comment (ashleymichal) [reply src/commands/build/wranglerjs/metadata.rs]: just to update, @cwndrws clarified that yes that's exactly what he was saying.

**Comment by xtuc** [comment]
I made an update, mainly based on https://github.com/cloudflare/wrangler/pull/215#issuecomment-505381101.

@cwndrws comments:
> I agree that the KV config should be in wrangler.toml. I like the idea of requiring all three values: (title, id, and binding) like @ashleymichal suggested.

The kv-namespace configuration is only possible in `wrangler.toml`; I removed the auto-creation feature (for now).

The config looks like the following:
```toml
[[kv-namespaces]]
name = "KV_NAMESPACE_A"
id = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
``` 

I'm not sure we should ask the user for the namespace name because we won't use it, the API doesn't support it. It would be more confusing if these names get out-of-sync and it would still work, until we add a check which would be a breaking change.

> I think requiring title be a valid JS symbol would bite us eventually.

I like that idea actually, I updated https://github.com/cloudflare/wrangler/issues/90 and we can discuss there.

**Comment by cwndrws** [comment]
@xtuc 

> I'm not sure we should ask the user for the namespace name because we won't use it, the API doesn't support it. It would be more confusing if these names get out-of-sync and it would still work, until we add a check which would be a breaking change.

I agree with not using the namespace name. 

Your example config:
```toml
[[kv-namespaces]]
name = "KV_NAMESPACE_A"
id = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```
doesn't contain a binding symbol. Should it be:
```toml
[[kv-namespaces]]
name = "KV_NAMESPACE_A"
binding = "KV_NAMESPACE_A"
id = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```
Or something similar?

**Comment by xtuc** [comment]
@cwndrws the name is the binding, I'm happy to rename it if it's confusing to you.

**Comment by xtuc** [comment]
Blocked by #285

**Comment by ashleymichal** [comment]
> @cwndrws the name is the binding, I'm happy to rename it if it's confusing to you.

Please reference the design document:
https://docs.google.com/a/cloudflare.com/document/d/1Pe2xymJ5dvyNhFB5eJszqTGq70t3pemYm85YEjkEe44/edit?disco=AAAADNBj3nY

If there are disagreements on this syntax, please add them there.

**Comment by ashleymichal** [comment]
upon further reflection, i actually do think that long term this is the right way to specify these namespaces. the wrangler.toml for a workers project should regard kv namespaces as an external dependency, and the configuration of that dependency is a separate issue.

**Comment by xtuc** [comment]
I noticed that when the namespace id is wrong, the publishing API returns a unclear error (generic 400, workers.api.error.internal_server); it would be great to change it.

@ashleymichal I did the updated.

**Comment by xtuc** [comment]
Merging into feat-better-kv

**Comment by juancampuzano** [comment]
Is it possible to make kv-namespaces in a production project that uses wrangler or is it a feature in development?

Currently I do not see the documentation of this in the master.

**Pull request merged**